File: sdk/ios/Sources/Afftok/ConversionTracker.swift
import Foundation
import CommonCrypto

/// Conversion tracking implementation for AffTok SDK
public class ConversionTracker {
    
    private let options: AfftokOptions
    private let queue: OfflineQueue
    private let session: URLSession
    
    public init(options: AfftokOptions, queue: OfflineQueue) {
        self.options = options
        self.queue = queue
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = AfftokConfig.connectionTimeout
        config.timeoutIntervalForResource = AfftokConfig.readTimeout
        self.session = URLSession(configuration: config)
    }
    
    /// Track a conversion event
    public func trackConversion(params: ConversionParams) async -> AfftokResponse {
        let payload = buildConversionPayload(params: params)
        
        do {
            let response = try await sendRequest(endpoint: AfftokConfig.conversionEndpoint, payload: payload)
            if response.success {
                log("Conversion tracked successfully: \(params.transactionId)")
            } else {
                queue.enqueue(type: "conversion", payload: payload)
                log("Conversion queued for retry: \(params.transactionId)")
            }
            return response
        } catch {
            queue.enqueue(type: "conversion", payload: payload)
            log("Conversion queued (offline): \(params.transactionId), error: \(error.localizedDescription)")
            return AfftokResponse(
                success: false,
                message: "Conversion queued for offline retry",
                error: error.localizedDescription
            )
        }
    }
    
    /// Track conversion with additional metadata
    public func trackConversionWithMeta(params: ConversionParams, metadata: [String: Any]) async -> AfftokResponse {
        var payload = buildConversionPayload(params: params)
        payload["metadata"] = metadata
        
        do {
            let response = try await sendRequest(endpoint: AfftokConfig.conversionEndpoint, payload: payload)
            if !response.success {
                queue.enqueue(type: "conversion", payload: payload)
            }
            return response
        } catch {
            queue.enqueue(type: "conversion", payload: payload)
            return AfftokResponse(
                success: false,
                message: "Conversion with metadata queued for retry",
                error: error.localizedDescription
            )
        }
    }
    
    /// Build conversion payload
    private func buildConversionPayload(params: ConversionParams) -> [String: Any] {
        let timestamp = Int64(Date().timeIntervalSince1970 * 1000)
        let nonce = generateNonce()
        let deviceInfo = DeviceInfo.getDeviceInfo()
        
        var payload: [String: Any] = [
            "api_key": options.apiKey,
            "advertiser_id": options.advertiserId,
            "offer_id": params.offerId,
            "transaction_id": params.transactionId,
            "status": params.status,
            "currency": params.currency,
            "timestamp": timestamp,
            "nonce": nonce,
            "device_info": deviceInfo
        ]
        
        if let userId = options.userId {
            payload["user_id"] = userId
        }
        if let clickId = params.clickId {
            payload["click_id"] = clickId
        }
        if let amount = params.amount {
            payload["amount"] = amount
        }
        if let customParams = params.customParams {
            payload["custom_params"] = customParams
        }
        
        // Add signature
        payload["signature"] = generateSignature(timestamp: timestamp, nonce: nonce)
        
        return payload
    }
    
    /// Send HTTP request
    private func sendRequest(endpoint: String, payload: [String: Any]) async throws -> AfftokResponse {
        guard let url = URL(string: "\(options.baseUrl)\(endpoint)") else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(options.apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue(AfftokConfig.sdkVersion, forHTTPHeaderField: "X-SDK-Version")
        request.setValue(AfftokConfig.sdkPlatform, forHTTPHeaderField: "X-SDK-Platform")
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        
        if httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 {
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            return AfftokResponse(
                success: true,
                message: json?["message"] as? String,
                data: json
            )
        } else {
            // Try fallback endpoint
            if endpoint == AfftokConfig.conversionEndpoint {
                return try await sendFallbackRequest(endpoint: AfftokConfig.fallbackConversionEndpoint, payload: payload)
            }
            return AfftokResponse(
                success: false,
                error: "HTTP \(httpResponse.statusCode)"
            )
        }
    }
    
    /// Send request to fallback endpoint
    private func sendFallbackRequest(endpoint: String, payload: [String: Any]) async throws -> AfftokResponse {
        guard let url = URL(string: "\(options.baseUrl)\(endpoint)") else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(options.apiKey, forHTTPHeaderField: "X-API-Key")
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (_, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        
        if httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 {
            return AfftokResponse(success: true, message: "Conversion tracked via fallback")
        } else {
            return AfftokResponse(success: false, error: "Fallback failed: \(httpResponse.statusCode)")
        }
    }
    
    /// Generate HMAC-SHA256 signature
    private func generateSignature(timestamp: Int64, nonce: String) -> String {
        let dataToSign = "\(options.apiKey)|\(options.advertiserId)|\(timestamp)|\(nonce)"
        
        guard let keyData = options.apiKey.data(using: .utf8),
              let messageData = dataToSign.data(using: .utf8) else {
            return ""
        }
        
        var hmac = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        
        keyData.withUnsafeBytes { keyBytes in
            messageData.withUnsafeBytes { messageBytes in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256),
                       keyBytes.baseAddress, keyData.count,
                       messageBytes.baseAddress, messageData.count,
                       &hmac)
            }
        }
        
        return hmac.map { String(format: "%02x", $0) }.joined()
    }
    
    /// Generate random nonce
    private func generateNonce() -> String {
        let chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        return String((0..<32).map { _ in chars.randomElement()! })
    }
    
    private func log(_ message: String) {
        if options.debug {
            print("[AffTok Conversion] \(message)")
        }
    }
}

File: sdk/android/afftok-sdk/src/main/java/com/afftok/ConversionTracker.kt
package com.afftok

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Conversion tracking implementation for AffTok SDK
 */
class ConversionTracker(
    private val options: AfftokOptions,
    private val fingerprintProvider: FingerprintProvider,
    private val queue: OfflineQueue
) {
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(Config.CONNECTION_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .readTimeout(Config.READ_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .writeTimeout(Config.WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .build()
    
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    
    /**
     * Track a conversion event
     */
    suspend fun trackConversion(params: ConversionParams): AfftokResponse {
        val payload = buildConversionPayload(params)
        
        return withContext(Dispatchers.IO) {
            try {
                val response = sendRequest(Config.CONVERSION_ENDPOINT, payload)
                if (response.success) {
                    log("Conversion tracked successfully: ${params.transactionId}")
                } else {
                    // Queue for retry
                    queue.enqueue("conversion", payload)
                    log("Conversion queued for retry: ${params.transactionId}")
                }
                response
            } catch (e: Exception) {
                // Queue for offline retry
                queue.enqueue("conversion", payload)
                log("Conversion queued (offline): ${params.transactionId}, error: ${e.message}")
                AfftokResponse(
                    success = false,
                    error = e.message,
                    message = "Conversion queued for offline retry"
                )
            }
        }
    }
    
    /**
     * Track conversion with additional metadata
     */
    suspend fun trackConversionWithMeta(
        params: ConversionParams,
        metadata: Map<String, Any>
    ): AfftokResponse {
        val payload = buildConversionPayload(params).toMutableMap()
        payload["metadata"] = metadata
        
        return withContext(Dispatchers.IO) {
            try {
                val response = sendRequest(Config.CONVERSION_ENDPOINT, payload)
                if (!response.success) {
                    queue.enqueue("conversion", payload)
                }
                response
            } catch (e: Exception) {
                queue.enqueue("conversion", payload)
                AfftokResponse(
                    success = false,
                    error = e.message,
                    message = "Conversion with metadata queued for retry"
                )
            }
        }
    }
    
    /**
     * Build conversion payload
     */
    private fun buildConversionPayload(params: ConversionParams): Map<String, Any> {
        val timestamp = System.currentTimeMillis()
        val nonce = generateNonce()
        val deviceInfo = fingerprintProvider.getDeviceInfo()
        
        val payload = mutableMapOf<String, Any>(
            "api_key" to options.apiKey,
            "advertiser_id" to options.advertiserId,
            "offer_id" to params.offerId,
            "transaction_id" to params.transactionId,
            "status" to params.status,
            "currency" to params.currency,
            "timestamp" to timestamp,
            "nonce" to nonce,
            "device_info" to deviceInfo
        )
        
        options.userId?.let { payload["user_id"] = it }
        params.clickId?.let { payload["click_id"] = it }
        params.amount?.let { payload["amount"] = it }
        params.customParams?.let { payload["custom_params"] = it }
        
        // Add signature
        payload["signature"] = generateSignature(payload, timestamp, nonce)
        
        return payload
    }
    
    /**
     * Send HTTP request
     */
    private fun sendRequest(endpoint: String, payload: Map<String, Any>): AfftokResponse {
        val url = "${options.baseUrl}$endpoint"
        val json = JSONObject(payload).toString()
        val body = json.toRequestBody(jsonMediaType)
        
        val request = Request.Builder()
            .url(url)
            .post(body)
            .addHeader("Content-Type", "application/json")
            .addHeader("X-API-Key", options.apiKey)
            .addHeader("X-SDK-Version", Config.SDK_VERSION)
            .addHeader("X-SDK-Platform", Config.SDK_PLATFORM)
            .build()
        
        val response = client.newCall(request).execute()
        val responseBody = response.body?.string()
        
        return if (response.isSuccessful) {
            val jsonResponse = responseBody?.let { JSONObject(it) }
            AfftokResponse(
                success = true,
                message = jsonResponse?.optString("message"),
                data = jsonResponse?.let { parseJsonToMap(it) }
            )
        } else {
            // Try fallback endpoint
            if (endpoint == Config.CONVERSION_ENDPOINT) {
                return sendFallbackRequest(Config.FALLBACK_CONVERSION_ENDPOINT, payload)
            }
            AfftokResponse(
                success = false,
                error = "HTTP ${response.code}: $responseBody"
            )
        }
    }
    
    /**
     * Send request to fallback endpoint
     */
    private fun sendFallbackRequest(endpoint: String, payload: Map<String, Any>): AfftokResponse {
        return try {
            val url = "${options.baseUrl}$endpoint"
            val json = JSONObject(payload).toString()
            val body = json.toRequestBody(jsonMediaType)
            
            val request = Request.Builder()
                .url(url)
                .post(body)
                .addHeader("Content-Type", "application/json")
                .addHeader("X-API-Key", options.apiKey)
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                AfftokResponse(success = true, message = "Conversion tracked via fallback")
            } else {
                AfftokResponse(success = false, error = "Fallback failed: ${response.code}")
            }
        } catch (e: Exception) {
            AfftokResponse(success = false, error = "Fallback error: ${e.message}")
        }
    }
    
    /**
     * Generate HMAC-SHA256 signature
     */
    private fun generateSignature(payload: Map<String, Any>, timestamp: Long, nonce: String): String {
        val dataToSign = "${options.apiKey}|${options.advertiserId}|$timestamp|$nonce"
        
        return try {
            val mac = Mac.getInstance("HmacSHA256")
            val secretKey = SecretKeySpec(options.apiKey.toByteArray(), "HmacSHA256")
            mac.init(secretKey)
            val hmacBytes = mac.doFinal(dataToSign.toByteArray())
            hmacBytes.joinToString("") { "%02x".format(it) }
        } catch (e: Exception) {
            log("Signature generation failed: ${e.message}")
            ""
        }
    }
    
    /**
     * Generate random nonce
     */
    private fun generateNonce(): String {
        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        return (1..32).map { chars.random() }.joinToString("")
    }
    
    /**
     * Parse JSON to Map
     */
    private fun parseJsonToMap(json: JSONObject): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        json.keys().forEach { key ->
            map[key] = json.get(key)
        }
        return map
    }
    
    private fun log(message: String) {
        if (options.debug) {
            println("[AffTok Conversion] $message")
        }
    }
}

